mapgen/iwyu-mapgen-std-symbol.py

```diff
@@ -0,0 +1,564 @@
+
+            # Look up names and qualify them with missing namespaces. Exclude
+            # names that are preceded by '::' (they are already qualified).
+            p = re.compile(r'(?<!::)\b\w+\b')
+            m = p.search(tpl_args)
+            while m:
+                arg_id = self.look_up(m.group(0))
+                tpl_args = tpl_args[:m.start()] + arg_id + tpl_args[m.end():]
+                m = p.search(tpl_args, m.start() + len(arg_id))
+
+            # Format pointer, reference, and function types as IWYU expects
+            # (i.e., add a space).
+            tpl_args = re.sub(r'(\w)([\*&(])', r'\1 \2', tpl_args)
+
+            id += tpl_args
+            self.skip_spaces()
+        return id
+
+    def look_up(self, name):
+        l = len(self.__namespace_stack)
+        # Look up from the nearest enclosing namespace.
+        for i in range(l):
+            id = '::'.join(self.__namespace_stack[: l-i]) + f'::{name}'
+            if self.found_names.count(id):
+                return id
+        return name
+
+    def skip_function(self):
+        t = self.__text
+        assert t[self.__cur_pos] == '('
+        self.skip_braced_text('(', ')')
+        self.skip_spaces()
+        while t[self.__cur_pos] not in '{;':
+            self.__cur_pos += 1
+        if t[self.__cur_pos] == ';':
+            self.__cur_pos += 1
+        elif t[self.__cur_pos] == '{':
+            self.skip_braced_text('{', '}')
+
+    def add_identifier_from_ns(self, id):
+        if self.__namespace_stack:
+            id = '::'.join(self.__namespace_stack) + f'::{id}'
+        if '\\' in id:
+            # Avoid names containing TeX commands.
+            return
+        if id not in self.found_names:
+            # Work around a bug in the standard text.
+            id = ERRONEOUS_SYMBOLS.get(id, id)
+            self.found_names.append(id)
+
+def process_synopsis(syn):
+    p = Parser(syn)
+    p.parse()
+    return p.found_names
+
+def process_tex_file(tex_file):
+    res = []
+    for m in SYNOPSIS.finditer(tex_file):
+        symbols = process_synopsis(m.group('code'))
+        res += zip(symbols, repeat(m.group('headername')))
+    return res
+
+def print_line(symbol, headername, lang):
+    if lang == 'c++':
+        print(f'{{ "{symbol}", kPrivate, "<{headername}>", kPublic }},')
+    else:
+        print(f'{{ "symbol": ["{symbol}", "private", '
+                              f'"<{headername}>", "public"] }},')
+
+def print_content(std_source_path, lang):
+    symbols_and_headernames = []
+    for path in sorted(glob.glob(os.path.join(std_source_path, '*.tex'))):
+        if path.endswith('iostreams.tex'):
+            # Skip until <iosfwd> can be handled correctly.
+            continue
+        with open(path, 'r') as f:
+            symbols_and_headernames += process_tex_file(f.read())
+
+    for symbol in sorted({item[0] for item in symbols_and_headernames}):
+        escaped = symbol.replace('"', r'\"')
```

Comment:
Could/should the escaping move into `print_line`?

Reply:
Ah, I see you would be re-escaping the symbol for every (symbol, header) pair if you do. I don't have a strong opinion either way.